### PR TITLE
fix(createClient): throw error on SK use

### LINKED ISSUE #330: Throw a descriptive error message if secret key is used instead of the public key

### PR DESCRIPTION
**Pull request by FlowFlorent**
Fixes #330

### CODE REVIEW

File: packages/liveblocks-client/src/client.ts [hunk 1]
```diff
@@ -181,6 +181,10 @@ function getThrottleDelayFromOptions(options: ClientOptions): number {
 function prepareAuthentication(clientOptions: ClientOptions): Authentication {
   // TODO: throw descriptive errors for invalid options
   if (typeof clientOptions.publicApiKey === "string") {
+    if (clientOptions.publicApiKey.startsWith("sk_"))
```
Comment (FlowFlorent) [inline packages/liveblocks-client/src/client.ts:184]: Wasn't sure wether to use `ifError` instead, let me know if it should be updated

**Comment by FlowFlorent** [comment]
There are two ways to go about fixing this problem: 
1. Exclusion -> if `sk_` then throw
2. Inclusion -> `if `pk_` then do not throw

I've implemented 1., let me know if this works 🙏

File: packages/liveblocks-client/src/client.ts (hunk 1 above)
Comment (nvie) [reply packages/liveblocks-client/src/client.ts:184]: No, this is fine I think. `ifError` will only error in development mode, but if you use an `sk_` key in production in a public key position, that will never work anyway, so better to catch it in the client, always, even (especially?) in production.

Comment (nvie) [reply packages/liveblocks-client/src/client.ts:184]: I'd probably catch both positive and negative cases here, aka structure it like:

```ts
const { publicApiKey, authEndpoint } = clientOptions;

if (publicApiKey !== undefined) {
  // A-ha! They're trying to use a public key! Let's check a few more things then…

  // First of all, ensure that they're not *also* setting the authEndpoint key!
  if (authEndpoint !== undefined) {
    // Throw friendly error explaining you cannot use both, and need to chose either of the two
  }

  if (typeof publicApiKey !== 'string') {
    // Throw friendly error that it should be a string, not whatever they used. Perhaps also hint that they should look for a key starting with `pk_` (and provide a link to the dashboard in there?)
  }

  if (startsWith('sk_')) {
    // A-ha! User provided the *secret* key! A common mistake! Give a friendly/helpful error message that they need to use the other one (and provide a link to the dashboard in there?)
  } else if (!publicApiKey.startsWith('pk_')) {
    // This is neither a public or a secret key, but another string. Tell them to look for their `pk_` key (and provide a link to the dashboard in there?)
  }

  // We made it! By now this is a (syntactically) valid PK, and we can continue
}
```

**Review by nvie — APPROVED** [review]
This is looking good, but I think we can  take the error messaging a step further and provide even more useful checks. Doing all these checks really [helps to not make our users think](https://docs.temporal.io/blog/error-message-design/) while setting up Liveblocks in their projects.

File: packages/liveblocks-client/src/client.ts (hunk 1 above)
Comment (FlowFlorent) [reply packages/liveblocks-client/src/client.ts:184]: Great comment, indeed, might as well make it bulletproof, all these use cases will end up happening at some point

Comment (FlowFlorent) [reply packages/liveblocks-client/src/client.ts]: @nvie I've addressed your comments, let me know if this works 🙏

File: packages/liveblocks-client/src/client.node.test.ts [hunk 2]
```diff
@@ -31,6 +31,55 @@ function createClientAndEnter(options: ClientOptions) {
...
+    ],
+    [
+      undefined,
+      null,
+      "authEndpoint must be a string or a function. For more information: https://liveblocks.io/docs/api-reference/liveblocks-client#createClientAuthEndpoint",
+    ],
+    [
+      "sk_xxx",
+      undefined,
+      "Invalid key. You are using the secret key which is not supported. Please use the public key instead. For more information: https://liveblocks.io/docs/api-reference/liveblocks-client#createClientPublicKey",
```
Comment (GuillaumeSalles) [inline packages/liveblocks-client/src/client.node.test.ts:54]: I would replace `key` by `publicApiKey` to give instant feedback on which option is used incorrectly.

```suggestion
      "Invalid publicApiKey. You are using the secret key which is not supported. Please use the public key instead. For more information: https://liveblocks.io/docs/api-reference/liveblocks-client#createClientPublicKey",
```

**Review by GuillaumeSalles — APPROVED** [review]
Awesome job! Really eager to see if it lowers the setup friction!

File: packages/liveblocks-client/src/client.node.test.ts (hunk 2 above)
Comment (FlowFlorent) [reply packages/liveblocks-client/src/client.node.test.ts:54]: Good improvement! Will adjust

**Pull request merged**